polkadot/network/src/lib.rs

```diff
@@ -420,10 +389,40 @@ impl PolkadotProtocol {
 			}
 			Message::BlockData(req_id, data) => self.on_block_data(ctx, peer_id, req_id, data),
 			Message::Collation(relay_parent, collation) => self.on_collation(ctx, peer_id, relay_parent, collation),
-			Message::CollatorRole(_) => {},
+			Message::CollatorRole(role) => self.on_new_role(ctx, peer_id, role),
 		}
 	}
 
+	fn on_session_key(&mut self, ctx: &mut Context<Block>, peer_id: PeerId, key: SessionKey) {
+		{
+			let info = match self.peers.get_mut(&peer_id) {
+				Some(peer) => peer,
+				None => return,
+			};
+
+			if !info.claimed_validator {
+				ctx.disable_peer(peer_id, "Session key broadcasted without setting authority role");
+				return;
```

Comment:
does the underlying protocol guarantee message order?

Because we do the same not-reactong `on_new_role` below, so if we receive messages out of order, we might have dropped the role-assignment for the peer and our disconnect is unjustified (and we could race that).

Reply:
`claimed_validator` is something that happens in the initial `Status` message, but I agree that if `Status` can race against `SessionKey` then something would break. I asked @tomaka about this and libp2p uses TCP connections that guarantee message order. Dep2p does also